File: LogicSim/LogicSim/Output.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace LogicSim
{
    class Output : IConnection
    {
        private Component _Component;
        private Link _Link;
        private bool _Powered;

        /// <summary>
        /// whether output is powered.
        /// </summary>
        public bool Powered {
            get
            {
                return _Powered;
            }
            set
            {
                if (_Powered != value)
                {
                    foreach(Input input in Link.Inputs)
                    {
                        Board.WriteBuffer[input.Component.Index] = true;
                    }

                    _Powered = value;
                }
            }
        }

        /// <summary>
        /// Component connected to the output.
        /// </summary>
        public Component Component
        {
            get
            {
                return _Component;
            }
            private set
            {
                _Component = value ?? throw new ArgumentNullException("Component");
            }
        }

        /// <summary>
        /// Link associated to the output.
        /// </summary>
        public Link Link
        {
            get
            {
                return _Link;
            }
            private set
            {
                _Link = value ?? throw new ArgumentNullException("Link");
            }
        }

        /// <summary>
        /// Creates a new empty output.
        /// </summary>
        public Output()
        {

        }

        /// <summary>
        /// Creates a new output.
        /// </summary>
        /// <param name="component">Component connected to the output</param>
        public Output(Component component)
        {
            Component = component;
        }

        /// <summary>
        /// Creates a new output.
        /// </summary>
        /// <param name="link">Link associated to the output</param>
        public Output(Link link)
        {
            Link = link;
        }

        /// <summary>
        /// Creates a new output.
        /// </summary>
        /// <param name="component">Component connected to the output</param>
        /// <param name="link">Link associated to the output</param>
        public Output(Component component, Link link)
        {
            Component = component;
            Link = link;
        }
    }
}

File: src/output.cpp
#include "output.h"
#include "link.h"
#include "input.h"
#include "board.h"
#include "component.h"

Output::Output(Component* component, Link* link)
{
	this->component = component;
	this->link = link;
	this->powered = false;
}

Output::~Output()
{
}

bool Output::getPowered()
{
	return powered;
}

void Output::setPowered(bool state) {
	if (state != powered) {
		powered = state;

		for (int i = 0; i < link->inputCount; i++)
			link->board->writeBuffer[link->inputs[i]->getComponent()->componentIndex] = true;
	}
}

Component* Output::getComponent()
{
	return component;
}

Link* Output::getLink()
{
	return link;
}